src/utils.js

```diff
@@ -157,17 +157,50 @@ export function parseGPTSingleSizeArray(singleSize) {
   }
 };
 
-exports.getTopWindowLocation = function () {
-  let location;
+export function getTopWindowLocation() {
+  if (exports.inIframe()) {
+    return getIframeParentLoc();
+  } else {
+    return window.location;
+  }
+}
+
+const getIframeParentLoc = function() {
+  let loc = window.location;
   try {
-    // force an exception in x-domain enviornments. #1509
-    window.top.location.toString();
-    location = window.top.location;
+    if (window.$sf) {
+      loc = window.document.referrer;
+    } else {
+      if (window.document.location && window.document.location.ancestorOrigins &&
+        window.document.location.ancestorOrigins.length >= 1) {
+        loc = window.document.location.ancestorOrigins[window.document.location.ancestorOrigins.length - 1];
+      } else if (window.document.location) {
+        loc = getNonWebKitIframeParentLoc();
+      }
+    }
+    loc = parseFullUrl(loc);
   } catch (e) {
-    location = window.location;
+    this.logMessage('getTopParentLoc failure', e);
+  }
+  return loc;
+};
+
+const getNonWebKitIframeParentLoc = function() {
+  let referrerLoc = '';
+  let currentWindow;
+  do {
+    currentWindow = currentWindow ? currentWindow.parent : window;
+    if (currentWindow.document && currentWindow.document.referrer) {
+      referrerLoc = currentWindow.document.referrer;
+    }
   }
+  while (currentWindow !== window.top);
+  return referrerLoc;
+};
 
-  return location;
+const parseFullUrl = function(locString) {
```

Comment:
There's a `parse` function in the `src/url.js` module that parses a url and returns a similar looking object, would reusing that function in place of this one fit this use case?

Reply:
Yes, though I added the `href` into the return object of that function.